File: baekjoon/Level/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.py
N = int(input())
A = list(map(int, input().split()))

def dp(array):
    cache = [1] * N
    for i in range(N):
        for j in range(i):
            if A[i] > A[j]:
                cache[i] = max(cache[i], cache[j] + 1)
    return cache

upCache = dp(A)
A.reverse()
downCache = dp(A)
downCache.reverse()

maxResult = 0
for i in range(N):
    maxResult = max(maxResult, upCache[i] + downCache[i] - 1)
print(maxResult)

File: baekjoon/Level/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.swift
let N = Int(readLine()!)!
var A = readLine()!.split(separator: " ").map{ Int($0)! }

func dp(array: [Int]) -> [Int] {
    var cache = Array(repeating: 1, count: N)
    for i in 0..<N {
        for j in 0..<i {
            if array[i] > array[j] {
                cache[i] = max(cache[i], cache[j] + 1)
            }
        }
    }
    return cache
}

var upCache = dp(array: A)
A.reverse()
var downCache = dp(array: A)
downCache.reverse()

var maxResult = 0
for i in 0..<N {
    maxResult = max(maxResult, upCache[i] + downCache[i] - 1)
}
print(maxResult)